ADALiOS/ADALiOS/ADTokenCache.m

```diff
@@ -0,0 +1,546 @@
+//  ADKeychainTokenCache.
+//
+//  The cache itself is a serialized collection of object and dictionaries in the following schema:
+//
+//  root
+//    |- version - a NSString with a number specify the version of the cache
+//    |- tokenCache - an NSDictionary
+//          |- tokens   - a NSDictionary containing all the tokens
+//          |     |- [<user_id> - an NSDictionary, keyed off of an NSString of the userId
+//          |            |- <ADTokenCacheStoreKey> - An ADTokenCacheItem, keyed with an ADTokenCacheStoreKey
+//          |- idtokens - An NSDictionary containing all of the idtokens, keyed off of the userID
+//
+
+#import "ADTokenCache.h"
+#import "ADAuthenticationError.h"
+#import "ADLogger+Internal.h"
+#import "ADErrorCodes.h"
+#import "ADTokenCacheKey.h"
+#import "ADTokenCacheItem.h"
+#import "ADUserInformation.h"
+#import "ADTokenCache+Internal.h"
+#import "ADTokenCacheKey.h"
+
+#define CHECK_ERROR(_cond, _code, _details) { \
+    if (!(_cond)) { \
+        ADAuthenticationError* _AD_ERROR = [ADAuthenticationError errorFromAuthenticationError:_code protocolCode:nil errorDetails:_details]; \
+        if (error) { *error = _AD_ERROR; } \
+        return NO; \
+    } \
+}
+
+@implementation ADTokenCache
+
+- (void)setDelegate:(nullable id<ADTokenCacheDelegate>)delegate
+{
+    _delegate = delegate;
+    _cache = nil;
+    
+    if (!delegate)
+    {
+        return;
+    }
+    
+    [_delegate willAccessCache:self];
+    
+    [_delegate didAccessCache:self];
+}
+
+- (nullable NSData *)serialize
+{
+    if (!_cache)
+    {
+        return nil;
+    }
+    
+    NSDictionary* wrapper = @{ @"version" : @CURRENT_WRAPPER_CACHE_VERSION,
+                               @"tokenCache" : _cache };
+    
+    @try
+    {
+        return [NSKeyedArchiver archivedDataWithRootObject:wrapper];
+    }
+    @catch (id exception)
+    {
+        // This should be exceedingly rare as all of the objects in the cache we placed there.
+        AD_LOG_ERROR(@"Failed to serialize the cache!", AD_ERROR_BAD_CACHE_FORMAT, nil, nil);
+        return nil;
+    }
+}
+
+- (id)unarchive:(NSData*)data
+          error:(ADAuthenticationError * __nullable __autoreleasing * __nullable)error
+{
+    @try
+    {
+        return [NSKeyedUnarchiver unarchiveObjectWithData:data];
+    }
+    @catch (id expection)
+    {
+        ADAuthenticationError* adError =
+        [ADAuthenticationError errorFromAuthenticationError:AD_ERROR_BAD_CACHE_FORMAT
+                                               protocolCode:nil
+                                               errorDetails:@"Failed to unarchive data blob from -deserialize!"];
+        
+        if (error)
+        {
+            *error = adError;
+        }
+        
+        return nil;
+    }
+}
+
+
+- (BOOL)deserialize:(nullable NSData*)data
+              error:(ADAuthenticationError * __nullable __autoreleasing * __nullable)error
+{
+    // Start by dropping our existing cache
+    _cache = nil;
```

Comment:
failure to deserialize should not result in cache state change. Last state of the cache should be preserved. so, override cache only if deserialize was successful.